h/admin/services/user.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+from h import models
+from h.api.search.index import BatchIndexer
+from h.util.user import userid_from_username
+
+
+class UserRenameError(Exception):
+    pass
+
+
+class RenameUserService(object):
+    """
+    Renames a user and updates all its annotations and when necessary the NIPSA
+    status.
+
+    ``check`` should be called first
+
+    Validates the new username and updates the User. The permissions of
+    the user's annotations are updated to reflect the new username.
+
+    May raise a ValueError if the new username does not validate or
+    UserRenameError if the new username is already taken by another account.
+    """
+    def __init__(self, request):
+        self.request = request
+        self.session = request.db
+
+    def check(self, old_username, new_username):
+        user = models.User.get_by_username(self.session, old_username)
+        if user is None:
+            raise UserRenameError('Could not find existing user with username "%s"' % old_username)
+
+        existing_user = models.User.get_by_username(self.session, new_username)
+        if existing_user:
+            raise UserRenameError('Another user already has the username "%s"' % new_username)
+
+    def rename(self, old_username, new_username):
```

Comment:
Perhaps the first argument here should be the user's integer ID rather than their username.

This is an extreme edge case, but in theory if three tasks to perform a rename were scheduled concurrently (A -> B, A -> C, D->A) this could end up resulting in (A -> B, D -> C) where D -> C is not one of the attempted renames.